src/cljs/vanilla/core.cljs

```diff
@@ -33,31 +36,39 @@
 
 (enable-console-print!)
 
-
-
-
-;;;;;;;;;;;;;;;;;;;;;
-;
-; HACK
-;
-; TODO - remove dummy "new-widget" stuff
-
-(def next-widget-idx (atom 0))
-
-(defn add-canned-widget [])
-  ;(let [rand-widget (get defs/widgets @next-widget-idx)]
-  ;
-  ;  ;(prn "add-canned-widget " rand-widget ", " @next-widget-idx, ", " (count defs/widgets))
-  ;
-  ;  (add-widget (grid/fixup-new-widget rand-widget))
-  ;  (if (< @next-widget-idx (dec (count defs/widgets)))
-  ;    (swap! next-widget-idx inc)
-  ;    (reset! next-widget-idx 0))))
-
-
-;;;;;;;;;;;;;;;;;;;;;
-;
-; END HACK
+(def conversion {:ret_types edn/read-string
+                 :key edn/read-string
+                 :name edn/read-string
+                 :basis edn/read-string
+                 :data-grid edn/read-string
+                 :type edn/read-string
+                 :data-source edn/read-string
+                 :options edn/read-string})
+
+(rf/reg-event-db
+  :set-layout
+  (fn-traced [db [_ layout-data]]
+   ;(prn "Set-layout start: " layout-data)
+
+     (let [data (:layout layout-data)
+           converted-data (mapv (fn [{:keys [ret_types key name basis data-grid type data-source options] :as original}]
+                                  (assoc original
+                                    :ret_types ((:ret_types conversion) ret_types)
+                                    :key ((:key conversion) key)
+                                    :name ((:name conversion) name)
+                                    :basis ((:basis conversion) basis)
+                                    :data-grid ((:data-grid conversion) data-grid)
+                                    :type ((:type conversion) type)
+                                    :data-source ((:data-source conversion) data-source)
+                                    :options ((:options conversion) options)))
+                                data)
+           highestNextid (apply max (mapv #(:key %) converted-data))]
```

Comment:
clever, but lets thing about possible error conditions.

Reply:
Agreed, we'll need to figure out a long term solution for the widget id.  Since querying records will be username based, it shouldnt really matter what the widget id is, as long as its unique.  So probably implement some kind of generated random one?

Reply:
have a look at [https://github.com/lbradstreet/cljs-uuid-utils](https://github.com/lbradstreet/cljs-uuid-utils)

Reply:
we can add this in as part of adding the :user-id token to the layout updates, to keep each user's UI separate.